crates/weaver_forge/src/file_loader.rs

```diff
@@ -0,0 +1,357 @@
+/// can be overridden by the user.
+pub struct EmbeddedFileLoader {
+    target: String,
+    embedded_dir: &'static include_dir::Dir<'static>,
+    fs_loader: Option<FileSystemFileLoader>,
+}
+
+impl EmbeddedFileLoader {
+    /// Create a new embedded file loader.
+    ///
+    /// If the `local_dir/target` directory exists, the loader will use the file system loader.
+    /// Otherwise, it will use the embedded directory.
+    pub fn try_new(
+        embedded_dir: &'static include_dir::Dir<'static>,
+        local_dir: PathBuf,
+        target: &str,
+    ) -> Result<Self, Error> {
+        let target_embedded_dir = embedded_dir.get_dir(target);
+        let target_local_dir = local_dir.join(target);
+        if let Some(dir) = target_embedded_dir {
+            Ok(Self {
+                target: target.to_owned(),
+                embedded_dir: dir,
+                fs_loader: if target_local_dir.exists() {
+                    Some(FileSystemFileLoader::try_new(local_dir, target)?)
+                } else {
+                    None
+                },
+            })
+        } else {
+            Err(TargetNotSupported {
+                root_path: embedded_dir.path().to_string_lossy().to_string(),
+                target: target.to_owned(),
+                error: "Target not found".to_owned(),
+            })
+        }
+    }
+}
+
+impl FileLoader for EmbeddedFileLoader {
+    /// Returns a textual representation of the root path of the loader.
+    /// This representation is mostly used for debugging and logging purposes.
+    fn root(&self) -> &Path {
+        if let Some(fs_loader) = &self.fs_loader {
+            return fs_loader.root();
+        }
+        self.embedded_dir.path()
+    }
+
+    /// Returns a list of all files in the loader's root directory.
+    fn all_files(&self) -> Vec<PathBuf> {
+        fn collect_files<'a>(dir: &'a include_dir::Dir<'a>, paths: &mut Vec<PathBuf>) {
+            for entry in dir.entries() {
+                match entry {
+                    include_dir::DirEntry::Dir(d) => collect_files(d, paths),
+                    include_dir::DirEntry::File(f) => {
+                        let relative_path = f.path().strip_prefix(dir.path()).expect("Failed to strip prefix. Should never happen as `dir.path()` is initial root.");
+                        paths.push(relative_path.to_owned());
+                    }
+                }
+            }
+        }
+
+        if let Some(fs_loader) = &self.fs_loader {
+            return fs_loader.all_files();
+        }
+
+        let mut files = vec![];
+        collect_files(self.embedded_dir, &mut files);
+        files
+    }
+
+    /// Returns a function that loads a file from a given name
+    fn file_loader(
+        &self,
+    ) -> Arc<dyn for<'a> Fn(&'a str) -> Result<Option<String>, Error> + Send + Sync + 'static> {
+        if let Some(fs_loader) = &self.fs_loader {
```

Comment:
Should the presence of a file system load completely OVERRIDE looking up a file, or should we *try* to look up using the fs_loader and on failure look up from the embeded target?

Reply:
It seems simpler to have an all-or-nothing logic in this specific case. A more granular approach risks involving more complicated logic when, for example, we want to remove a template defined in the embedded version but that should not be in the overloaded version.

Reply:
AH - since you're exposing all_files, I think I agree.  I wasn't seeing the full design before.